qcodes/tests/dataset/test_database_extract_runs.py

```diff
@@ -411,17 +419,190 @@ def test_load_by_X_functions(two_empty_temp_db_connections,
         ds.mark_completed()
 
     extract_runs_into_db(source_path, target_path, source_ds_2_2.run_id)
+    extract_runs_into_db(source_path, target_path, source_ds_2_1.run_id)
+    extract_runs_into_db(source_path, target_path, source_ds_1_1.run_id)
 
     test_ds = load_by_guid(source_ds_2_2.guid, target_conn)
     assert source_ds_2_2.the_same_dataset_as(test_ds)
 
     test_ds = load_by_id(1, target_conn)
     assert source_ds_2_2.the_same_dataset_as(test_ds)
 
-    test_ds = load_by_counter(1, 1, target_conn)
+    test_ds = load_by_run_spec(captured_run_id=source_ds_2_2.captured_run_id,
+                               conn=target_conn)
+    assert source_ds_2_2.the_same_dataset_as(test_ds)
+
+    assert source_exp2.exp_id == 2
+
+    # this is now the first run in the db so run_id is 1
+    target_run_id = 1
+    # and the experiment ids will be interchanged.
+    target_exp_id = 1
+
+    test_ds = load_by_counter(target_run_id, target_exp_id, target_conn)
     assert source_ds_2_2.the_same_dataset_as(test_ds)
 
 
+def test_combine_runs(two_empty_temp_db_connections,
+                      empty_temp_db_connection,
+                      some_interdeps):
+    """
+    Test that datasets that are exported in random order from 2 datasets
+    can be reloaded by the original captured_run_id and the experiment
+    name.
+    """
+    source_conn_1, source_conn_2 = two_empty_temp_db_connections
+    target_conn = empty_temp_db_connection
+
+    source_1_exp = Experiment(conn=source_conn_1,
+                              name='exp1',
+                              sample_name='no_sample')
+    source_1_datasets = [DataSet(conn=source_conn_1,
+                                 exp_id=source_1_exp.exp_id) for i in range(10)]
+
+    source_2_exp = Experiment(conn=source_conn_2,
+                              name='exp2',
+                              sample_name='no_sample')
+
+    source_2_datasets = [DataSet(conn=source_conn_2,
+                                 exp_id=source_2_exp.exp_id) for i in range(10)]
+
+    source_all_datasets = source_1_datasets + source_2_datasets
+
+    shuffled_datasets = source_all_datasets.copy()
+    random.shuffle(shuffled_datasets)
```

Comment:
I have a weak opinion that we should avoid randomness inside tests. I can take on a (probably cumbersome) rewrite of this logic with itertools and/or hypothesis.

Reply:
Sure if you like. I could not be bothered in this case